File: Day4/Longest-Subarray-with-Sum-K/Method1.cpp
class Solution
{
public:
    int longestSubarray(vector<int> &arr, int k)
    {
        // code here
        unordered_map<int, int> prefix;
        int maxlength = 0, sum = 0, n = arr.size();

        for (int i = 0; i < n; i++)
        {
            sum += arr[i];
            if (sum == k)
                maxlength = max(maxlength, i + 1);
            if (prefix.find(sum - k) != prefix.end())
            {
                int len = i - prefix[sum - k];
                maxlength = max(maxlength, len);
            }

            if (prefix.find(sum) == prefix.end())
                prefix[sum] = i;
        }

        return maxlength;
    }
};

File: Day4/Longest-Subarray-with-Sum-K/Method1.java
// User function Template for Java

class Solution {
    public int longestSubarray(int[] arr, int k) {
        HashMap<Integer, Integer> hs = new HashMap<>();

        int n = arr.length, maxLen = 0, sum = 0;

        for (int i = 0; i < n; i++) {
            sum += arr[i];
            if (sum == k)
                maxLen = Math.max(maxLen, i + 1);
            if (hs.containsKey(sum - k)) {
                int len = i - hs.get(sum - k);
                maxLen = Math.max(maxLen, len);
            }
            if (!hs.containsKey(sum)) {
                hs.put(sum, i);
            }
        }
        return maxLen;
    }
}
